### PR TITLE
Firebase configuration

### LINKED ISSUE #1: 1. Como un componente, quiero ser capaz de leer y escribir en la base de datos de Firestore, así que los usuarios puedan guardar su información.
Esta es 1 de 2 historias que deben completarse en la primera semana para permitir un mayor desarrollo de la aplicación.

AC:

* Se han añadido: `firebase` & `react-firebase-hooks` como dependencias del proyecto.
* Realice un cambio en la base de datos de Firestore y este aparecerá en la aplicación.
* Realice un cambio en el app y este aparecerá en la base de datos de Firestore.

Notas:

* No realice mucho aquí. En el pasado, los pares han construido todo el formulario para adicionar un artículo en la primera semana. Lo mejor es hacer una vista con un `<button>` y un `<div>`. Click el `<button>` para enviar datos a la base de datos. Que los datos aparezcan en el `<div/>`. ¡Simple!
* También, si tú necesitas más dirección sobre cómo usar `react-firebase-hooks` tú puedes consultar el commit de Alejandro: https://github.com/alejandronanez/smart-shopping-list-demo/commit/d590adc3224872e71d000c41977a1b7a84542c31 y/o ver el video de Loom (Idioma: English) donde él explica como esto funciona: https://www.loom.com/share/0dffa1a63bf24b449d6565c76606bbd9

### PR DESCRIPTION
**Pull request by yanurd**
## Description
This PR includes **_two components_** , **_one custom hook_** and **_two functions_**:

**Components:**

1. `CreateItem`: it is a button that creates a new item by a given default name for testing purposes
2. `ListItem`: displays a `<li>` with the items created. This component fetches Firebase API to retrieve them

**Custom Hook:**

1. `useProducts`: it is responsible for fetch data from Firebase and listen data changes

**Functions**

1. `helpers/firebaseHelpers.js`: we have a function that receives the results obtained in Firebase and parses the data, leaving them ready to be displayed on screen
2. `utils/firebaseUtils.js`: we have a function responsible for creating products in the Firebase

## Related Issue

Closes #1 

## Acceptance Criteria

- [x] Se han añadido: firebase & _react-firebase-hooks*_ como dependencias del proyecto.
- [x] Realice un cambio en la base de datos de Firestore y este aparecerá en la aplicación.
- [x] Realice un cambio en el app y este aparecerá en la base de datos de Firestore.

*It was recommended not to use _react-firebase-hooks_ as it breaks code, that is why we added custom hook.
## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|  ✓  | :link: Update dependencies |
|    | :scroll: Docs              |

## Gif

collab-gif

## Testing Steps / QA Criteria

1. From your terminal, pull down this branch with `git pull origin il-yu-config-firebase` and check that branch out with `git checkout il-yu-config-firebase`
2. Then `npm install` the newly added dependencies locally and `npm start` to launch the app.
3. Click the button to add a new item, it gives 'A new item' name by default. Make sure it adds it to the list and make sure that item appears in the items collection of the database [here](https://console.firebase.google.com/u/0/project/tcl-34-smart-shopping-list/firestore/data/~2Fproducts).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
   "private": true,
   "dependencies": {
     "@the-collab-lab/shopping-list-utils": "^1.0.1",
+    "firebase": "^9.1.1",
```
Comment (alxmcr) [inline package.json:7]: Great! You're using the Firebase latest version. 🎉

File: src/App.js [hunk 2]
```diff
@@ -1,24 +1,13 @@
...
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
-      </header>
+      <CreateItem />
+      <ListItem />
```
Comment (alxmcr) [inline src/App.js:10]: I like the simplicity and clarity in App.js 💚

File: src/components/ListItem.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { useProducts } from '../hooks/useProducts';
+
+const ListItem = () => {
+  const { products, loading, error } = useProducts();
+
+  return (
+    <div>
+      {loading && 'Loading'}
```
Comment (alxmcr) [inline src/components/ListItem.js:8]: In general, a comparison { something && result } is OK. However, there are some cases where you can get unexpected results.

For example: if an error is null, according to Javascript: the result could be falsy, then React renders nothing. In other cases, when loading is 1 (true) or 0 (false) in some cases depends on your condition.

Some people suggested that it's much better you use an explicit condition like:
{ loading ? <p>'Loading'... </p>: null }
{error !== null ? <p>There was an error.</p>: null} or {error !== null ? <p>{error.message}</p>: null}

**^^They are only suggestions, not blocking. 😃**

File: src/hooks/useProducts.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import { db } from '../lib/firebase';
+import { collection, query, onSnapshot, addDoc } from '@firebase/firestore';
+import { useState, useEffect } from 'react';
+
+export const useProducts = () => {
+  const [products, setProducts] = useState([]);
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState(false);
+
+  const addProduct = async () => {
```
Comment (alxmcr) [inline src/hooks/useProducts.js:10]: It's interesting 😮!  You have an "extra" function in your custom hook: `addProduct`. 

This works! But, I would like to ask you: 

❓ Did you consider moving your `addProduct` function to **another file** like `helpers` or **isolate this logic** from your custom hook?

File: package.json (hunk 1 above)
Comment (mailliwmogollon) [reply package.json:7]: Congratulations on your work....!!!

File: src/hooks/useProducts.js (hunk 4 above)
Comment (IsabellaSoares) [reply src/hooks/useProducts.js:10]: Thanks! I understood all the points, but I still had a doubt. Why does the fetching data function need to be a custom hook? Couldn't it stay in `firebaseUtils` too? In other words, I wonder if it's a convention that the fetching data function is a custom hook while the other operations don't have to be?

Comment (alxmcr) [reply src/hooks/useProducts.js:10]: You're welcome! Good question as well. Thank you 💯 !

About your question: _"Why does the fetching data function need to be a custom hook?"_ I'm not sure, but I guess Yancarlo & you (@IsabellaSoares)  decided to create a custom hook 😄, and you believed this is the best way to solve this issue 🥳  .  it's great you have created it💯 !

I didn't find a rule or policy that it says if you need to fetch data, you should create a custom hook.  But I support custom hooks when I ask myself: 

- Do I need to use some results in multiple places? -> **Extract logic, and share it.**
- Do I need to have my own states? (Your current code: `loading` state is reused, but I understood is a state when your fetching data, you could lose some control)
- Is my component difficult to read, and maintain? (of course: vision about the future, but avoid over-engineering)
- Do I want to have more control over **my logic, and isolation goals**? (For example: when is `loading`, fetching data, and happened an `error`)

Also, 👀 if we pay attention to small details your custom hook could be divide into more parts: `firebaseOperations.js` (CRUD), `firebaseHelpers.js` (transform Firebase docs to JS objects, filters, remove elements, and so on), and your `useProducts.js` could be simpler, and other great things. Maybe, you can do in the next issues 🚀 😄.

**Bonus.** There are other external libraries that follow this **unofficial "pattern"** like React Router, Redux, and so on. I believe it has good acceptance in React community.

**Happy for your curiosity and interest!** 🥳 😄

Comment (IsabellaSoares) [reply src/hooks/useProducts.js:10]: Wow! This is very interesting! 👀 We hadn't thought of all these possibilities to resolve the issue. It's the first time we've both implemented a custom hook, but the idea of ​​separating responsibilities into different files is also pretty cool! Thank you for the help! We will evaluate the possibilities and make the necessary corrections! 🚀

File: src/components/CreateItem.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  const [isLoading, setIsLoading] = useState(false);
+  const [error, setError] = useState(null);
+
+  const handleClick = () => {
+    setIsLoading(true);
+
+    addProduct()
+      .then((res) => console.log(res))
+      .catch((error) => setError(error))
+      .finally(() => setIsLoading(false));
```
Comment (jandrade) [inline src/components/CreateItem.js:14]: Nice touch!

File: src/components/ListItem.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        date,
+      ).toLocaleString()}`}</li>
+    ));
+  };
+
+  return (
+    <div>
+      {loading ? <p>Loading...</p> : null}
+      {error ? <p>There was an error</p> : null}
+      {!loading ? showProducts() : null}
```
Comment (jandrade) [inline src/components/ListItem.js:22]: What do you think about simplifying this logic a bit so we only have to handle two conditions?

```suggestion
      {loading ? <p>Loading...</p> : showProducts()}
      {error ? <p>There was an error</p> : null}
```

File: src/components/ListItem.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  const showProducts = () => {
+    if (products.length === 0) {
+      return <p>You don't have any product in your list</p>;
+    }
+
+    return products.map(({ id, name, date }) => (
+      <li key={id}>{`Product: ${name} => Created at: ${new Date(
+        date,
+      ).toLocaleString()}`}</li>
+    ));
```
Comment (jandrade) [inline src/components/ListItem.js:15]: I'd suggest to wrap this with an `<ul>` element to have valid markup.

Currently the output would be:
```
<div>
  <li>...</li>
</div>
```

File: src/helpers/firebaseHelpers.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export const parseData = (querySnapshot) =>
```
Comment (jandrade) [inline src/helpers/firebaseHelpers.js:1]: I see that you have created also `firebaseUtils.js`.... what's the difference between both files?

File: src/helpers/firebaseHelpers.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+  querySnapshot.docs.reduce((acc, doc) => {
+    const { name, date } = doc.data();
+    const id = doc.id;
+    return [...acc, { id, name, date }];
```
Comment (jandrade) [inline src/helpers/firebaseHelpers.js:5]: Nit: This can also be simplified to:
```suggestion
    return [...acc, { id: doc.id, name, date }];
```

**Review by jandrade — APPROVED** [review]
Looks really good to me! Approving because the PR covers all the main tasks that you need to do according to the issue. Just left a couple of minor comments related to some small refactors that I'll leave for your consideration, and also a couple of questions to get more context on the implementation.

I'm not sure if you have this on your radar.... one thing to consider is that `CreateItem` and `ListItem` will go away after merging both PRs, because the other PR (#2) will include the changes that cover that logic. Something worth to define with the other team to see how you handle the merge process :).

Great work!

File: src/components/CreateItem.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import { useState } from 'react';
+import { addProduct } from '../utils/firebaseUtils';
+
+const CreateItem = () => {
+  const [isLoading, setIsLoading] = useState(false);
```
Comment (alxmcr) [inline src/components/CreateItem.js:5]: Now, I like it so much that `CreateItem` has its own states. 🥳

File: src/components/CreateItem.js (hunk 5 above)
Comment (alxmcr) [reply src/components/CreateItem.js:14]: Totally agree! 👍

File: src/hooks/useProducts.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    setLoading(true);
+
+    // Retrieves information from Firestore
+    const getProducts = async () => {
+      try {
+        const productsCollection = await collection(db, 'products');
+        const queryProducts = query(productsCollection);
+
+        unsubscribe = onSnapshot(queryProducts, (querySnapshot) => {
+          setProducts(parseData(querySnapshot));
```
Comment (alxmcr) [inline src/hooks/useProducts.js:22]: `parseData` function is magic! 😄

File: src/helpers/firebaseHelpers.js (hunk 8 above)
Comment (alxmcr) [reply src/helpers/firebaseHelpers.js:1]: Well, about naming files. I know it's difficult to choose the best name but in this specific case. I suggested creating those files: `firebaseHelpers.js` and `firebaseOperations.js`. Maybe, it's only a small confusion.

- `firebaseOperations.js`: this file could have functions about Firebase like `addProduct`, `readProduct`, `updateProduct`, and `deleteProduct`.
- `firebaseHelpers.js`: it has small or medium functions that help us to "process" Firebase information as parsing data, filtering, getProductsByName,  getProductsByCreatedAt, and so on.

I'm so sorry for the confusion.

**Review by alxmcr — COMMENTED** [review]
I can't give you more approvals! but 
> Thank you so much for your huge work and effort! I really appreciate them! 

✔️ ✔️

**Comment by IsabellaSoares** [comment]
Thanks @alxmcr and @jandrade for the review and all the tips and teachings! 🚀

**Comment by alxmcr** [comment]
Great, thanks! Please 🙏 , when you have a chance, feel free to merge your PR.🥳 🎉 !  -- 💯

**Pull request merged**